### ISSUE TITLE
JSON data type always parsed as binary

### ISSUE THREAD

**Issue by panki** [sidorares/node-mysql2#409]
I have a table with JSON column.

Column options: encoding UTF-8, collationutf8mb4_general_ci.
Table options: CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci.

Query this column always returns json object with broken character set. For example:

Real data stored in mysql: {"value": "Problem: Arbeitsaufträge - привет"}
Query result: {"value": "Problem: ArbeitsauftraÌge - Ð¿ÑÐ¸Ð²ÐµÑ"}

I've made a little research and it turns out that mysql returns characterSet: 63 (binary) for that column, but node-mysql handles it well, while node-mysql2 - not.

Field definition in result set:

```
{ catalog: 'def',
    schema: '...',
    name: 'metadata',
    orgName: 'metadata',
    table: 'products',
    orgTable: 'products',
    characterSet: 63,
    columnLength: 4294967295,
    columnType: 245,
    flags: 144,
    decimals: 0
}
```

**Comment by sidorares** [comment]
node-mysql always decodes input data as utf-8 regardless of actual encoding used

`characterSet: 63` is not correct as well IMO, this basically means server is telling you "I don't know what encoding data is"

mysql2 related code is here - https://github.com/sidorares/node-mysql2/blob/f76a614fc1d343d322fd7e1aafe2971daae3ff0a/lib/compile_text_parser.js#L145-L147

Not sure exactly what to do here, maybe we should ignore encoding in case of JSON column ( or assume binary means utf8)

**Comment by sidorares** [comment]
do you have `63` charset for json column only? If you add another text column would it be returned as `utf8mb4`?

**Comment by panki** [comment]
@sidorares sorry, i was wrong about node-mysql, it doesn't parse json columns and returns them as utf8 strings.

**Comment by panki** [comment]
Table:
```
CREATE TABLE `test` (
  `id` int(11) NOT NULL,
  `json_column` json DEFAULT NULL,
  `text_column` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```

Query:
```
connection.query('SELECT * FROM test', function(err, rows, fields) {
  if (err) throw err;

  console.log(rows);
  console.log(fields);
});
```

Resuls:
```
[ TextRow {
    id: 1,
    json_column: { foo: 'Ð¿ÑÐ¸Ð²ÐµÑ' },
    text_column: 'foo - привет' } ]
```

Field definitions:
```
[ { catalog: 'def',
    schema: 'test',
    name: 'id',
    orgName: 'id',
    table: 'test',
    orgTable: 'test',
    characterSet: 63,
    columnLength: 11,
    columnType: 3,
    flags: 20483,
    decimals: 0 },
  { catalog: 'def',
    schema: 'test',
    name: 'json_column',
    orgName: 'json_column',
    table: 'test',
    orgTable: 'test',
    characterSet: 63,
    columnLength: 4294967295,
    columnType: 245,
    flags: 144,
    decimals: 0 },
  { catalog: 'def',
    schema: 'test',
    name: 'text_column',
    orgName: 'text_column',
    table: 'test',
    orgTable: 'test',
    characterSet: 224,
    columnLength: 1020,
    columnType: 253,
    flags: 0,
    decimals: 0 } ]
```

**Comment by sidorares** [comment]
can you try to specify json column charset as well?

```sql
CREATE TABLE `test` (
  `id` int(11) NOT NULL,
  `json_column` json CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `text_column` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;

```

**Comment by panki** [comment]
actually, json columns handled here https://github.com/sidorares/node-mysql2/blob/f76a614fc1d343d322fd7e1aafe2971daae3ff0a/lib/compile_binary_parser.js#L138

**Comment by panki** [comment]
> can you try to specify json column charset as well?

No, mysql does not allow to do it, it only allows to specify COLLATE, but if i try i'm getting error: COLLATION 'utf8mb4_unicode_ci' is not valid for CHARACTER SET 'binary'

**Comment by sidorares** [comment]
https://tools.ietf.org/html/rfc7159#page-9

> JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32.  The default
   encoding is UTF-8, and JSON texts that are encoded in UTF-8 are
   interoperable in the sense that they will be read successfully by the
   maximum number of implementations; there are many implementations
   that cannot successfully read texts in other encodings (such as
   UTF-16 and UTF-32).

I guess we need to ignore server charset here and just assume it's utf8, and "binary" charset in response field type is mysql quirk/bug

**Comment by panki** [comment]
Should i prepare PR with this change?

**Comment by sidorares** [comment]
@panki would be great if you can do this. Note that there is "text protocol" and "binary protocol"

https://github.com/sidorares/node-mysql2/blob/f76a614fc1d343d322fd7e1aafe2971daae3ff0a/lib/compile_text_parser.js#L145-L147
https://github.com/sidorares/node-mysql2/blob/f76a614fc1d343d322fd7e1aafe2971daae3ff0a/lib/compile_text_parser.js#L145-L147

We need to replace `JSON.perse...` code to be

```js
return 'JSON.parse(packet.readLengthCodedString('utf8'));';
```

in both and add test with `conn.query()` and `conn.execute()` with some unicode inserted initially in temp table, ` { "face": "\uD83D\uDE02" }`,  `assert(rows[0].json_column.face === '😂')`

note that utf8 is encoding name, not charset name ( mysql utf8 charset correspond to cesu-8 encoding and utf8mb4 -> utf8 encoding )

**Comment by panki** [comment]
@sidorares thank you for explanation, I'll do it soon

**Comment by panki** [comment]
@sidorares i've created PR, but i forgot to use issue number to link it to this issue, could you fix it?

**Comment by sidorares** [comment]
> i forgot to use issue number to link it to this issue

just leaving #410 here should cross-link it

**Comment by sushantdhiman** [comment]
Fixed in #410 , will be released soon:)
